---
title: "ME_classification"
author: "Miguel Parra"
date: '2023-07-06'
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

---
title: "EE_ME_classification"
author: "Miguel Parra"
date: '2023-06-16'
output: html_document
---


```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

The idea of this notebook is to check if machine learning is able to differentiate patients that were exposed to famine during mid embryonic development (between months 3 and 6 of development) and control patients that were not exposed to phamine, using methylation data and genomic data (PRS and SNPs).

## 0. Set-up

```{r}
library ("tidyverse") 
library ("ggplot2")
library ("minfi")
library ("MUVR")
library ("doParallel")
library ("bigsnpr")
library ("bigstatsr")
library ("pROC")
library ("randomForest")
library ("caret")
library ("Rtsne")
library ("janitor")
```

Import data:

```{r}

#Metadata: ----
samples_metadata <- readRDS("OUTPUT/feature_selection/rds/samples_metadata.rds")


#Methylation data: ----
grset <- readRDS("OUTPUT/methylation_preprocessing/rdsDir/grset_normalized_noSNPs_noSex.Rds")
beta_values <- getBeta (grset)
mval <- getM(grset)


#Genomic data: ----
snp_array <- readRDS("OUTPUT/genomic_preprocessing/rdsDir/bigSNPr_prepro.rds")

G <- snp_fastImputeSimple(snp_array$genotypes, method = "mode") #Imputing the remaining missing values
maf <- snp_MAF(G)
G <- G[]


colnames(G) <- snp_array$map$marker.ID
rownames(G) <- snp_array$fam$family.ID


#PRS: ----
df_all_prs <- readRDS("OUTPUT/PRS/rds/all_prs_df.rds")
df_all_prs$SXS <- df_all_prs$IID %>% str_split(., "_1") %>% sapply (., "[[", 1)

#Matrix with PRS
mt_all_prs <- df_all_prs %>% filter (SXS %in% samples_metadata$SXS) %>%  dplyr::select (contains("Pt")) %>% as.matrix()
rownames (mt_all_prs) <-  df_all_prs %>% filter (SXS %in% samples_metadata$SXS) %>% pull (SXS) 
```

Select the samples that were only exposed during mid development to the famine (first 6 months) and the ones that were concived after the famine.

```{r}

samples_metadata_fil <- samples_metadata %>% filter (both_views == 1) %>% filter (Exposure == "Mid" | Timeframe_non == "Postfamine" )

```

Exposure   n   percent
      Mid  85 0.3761062
      Non 141 0.62389


```{r}
path_output <- "OUTPUT/ME_classification"
n_folds <- 5 #Number of crossvalidation folds
```

## 1.MUVR:

### 1.1 Methylation + SNPs

#### -Preprocessing

Selection of the methylation variables with highest variance

```{r}
beta_variance <- apply (beta_values, 1, var)

#Select only values with high variance (top 50%)
betas_fil <- beta_values[ beta_variance >  quantile (beta_variance, 0.50),]

betas_fil <- t(betas_fil[,samples_metadata_fil$id])

```

Selection of SNPs with high maf

```{r}

G_fil <- G[samples_metadata_fil$SXS, which (maf > 0.1)]

```

#### -Model

CV of the RF model. Feature selection is included in the model.

```{r}
#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = 5)

muvr_list <- list()
rf_list <- list()
x_test_topvar <- list()

t1 <- Sys.time()
for (i in 1:5) {
  
  y_train_cv <- samples_metadata_fil[-folds[[i]],]$target
  y_test_cv <- samples_metadata_fil[folds[[i]],]$target
  
  
  
  ##Genomic data: 
  #Based on previous G_fil2 and betas_fil2, that were filtered based on variance or maf
  
  G_fil_cv_train <- G_fil[ (samples_metadata_fil[-folds[[i]],] %>% pull (SXS)),]
  G_fil_cv_test <- G_fil[(samples_metadata_fil[folds[[i]],] %>% pull (SXS)), ]
  
  x_square <- function (j){
    chi <- chisq.test(j, y_train_cv)
    return (chi$p.val)
  }
  pval_xsqr <- apply (G_fil_cv_train,2, function(j){chisq.test(j, y_train_cv)$p.val})
  #Selection of the top 5000 snps that correlate with target. 
  G_fil_cv_train <-  G_fil_cv_train[, sort(pval_xsqr, index.return = TRUE)$ix[1:2000]]
  G_fil_cv_test <-  G_fil_cv_test[, sort(pval_xsqr, index.return = TRUE)$ix[1:2000]]
  
  ##Methylation data: 
  
  #Transpose de matrix and select train and test samples
  betas_fil_cv_train <- betas_fil[(samples_metadata_fil[-folds[[i]],] 
                                       %>% pull (id)),]
  betas_fil_cv_test <- betas_fil[(samples_metadata_fil[folds[[i]], ] 
                                      %>% pull (id)), ]
  
  # Select top variables with lower pval
  pval_betas <- apply (betas_fil_cv_train,
                       2,
                       function (j) { t.test (j ~ y_train_cv)$p.val})
  betas_fil_cv_train <- betas_fil_cv_train [, sort (pval_betas, index.return = TRUE)$ix[1:2000]]
  betas_fil_cv_test <- betas_fil_cv_test [, sort (pval_betas, index.return = TRUE)$ix[1:2000]]
  
  
  
  
  
  ###MUVR
  
  cl = makeCluster(10)
  doParallel::registerDoParallel(cl)
  
  model_muvr_cv <- MUVR( X = cbind (G_fil_cv_train, betas_fil_cv_train), 
                         Y = y_train_cv, 
                         nRep = 10, 
                         nOuter = 5, 
                         varRatio = 0.7, 
                         method = "RF", 
                         fitness = "AUROC", 
                         modReturn = TRUE)
  
  stopCluster(cl)
  
  #Selecting the most important features according to the
  top_var_cv <- getVIP(model_muvr_cv, model = "mid")$name
  
  
  # ###RF caret: 
  # 
  # 
  # cl <- makeCluster(5)
  # doParallel::registerDoParallel(cl)
  # 
  # tuneGrid <- expand.grid(.mtry = c(2, 8, 20, 50, 100, 150))
  # train_control <- trainControl(method = "cv",
  #                               number = 5,
  #                               verboseIter = TRUE,
  #                               returnResamp = "final",
  #                               classProbs = TRUE,
  #                               summaryFunction = twoClassSummary
  # )
  # rf_fit_cv <- caret::train(x = cbind(G_fil_cv_train, betas_fil_cv_train)[,top_var_cv],
  #                           make.names(y_train_cv),
  #                           method = "rf",
  #                           trControl = train_control, 
  #                           tuneGrid = tuneGrid, 
  #                           ntree = 300
  # )
  # stopCluster(cl)
  
  

  #return models
  muvr_list[[i]] <-  model_muvr_cv
  # rf_list[[i]] <- rf_fit_cv
  x_test_topvar[[i]] <- cbind(G_fil_cv_test, betas_fil_cv_test)[,top_var_cv]
  
  #Show time taken
  Sys.time() - t1
}

```

Save results
```{r}
saveRDS(list (muvr_list, x_test_topvar), file = paste(path_output, "rds","MUVR_Methy_SNPs.rds", sep = "/"))
```


load results
```{r}
list_results <- readRDS (paste(path_output, "rds","MUVR_Methy_SNPs.rds", sep = "/"))

muvr_list <- list_results[[1]]
x_test_topvar <- list_results[[2]]
```


Top variables

```{r}
top_variables_list <- lapply (x_test_topvar, colnames)

common_top_variables <- Reduce (intersect, top_variables_list )
print (common_top_variables)

sapply (top_variables_list, length) 

# #Percentage of genomic variables in top variables
# sapply (top_variables_list, function(x){round (length (x[grep("rs", x)]) * 100/length (x), 2)}) %>% 
# 
# #Number of genomic variables in top variables
# sapply (top_variables_list, function(x){length (x[grep("rs", x)]) }) 


#Variables that appear in 3 of the folds
common_variables <- unique (unlist (top_variables_list))
common_top_variables <-  common_variables[sapply 
                                          ( common_variables,
                                            function(x){
                                              sum (sapply (top_variables_list, function (y) x %in% y))}) >= 4]

print (common_top_variables)
print (top_variables_list)

```
No genomic variables selected as top features. (Only methylation variables). 
3 variable selected in the 5 folds. ("cg08582801" "cg25669494" "cg02997497"). 2 other in 4 of the folds. 



#### -Plots

Roc plots of the MUVR models

```{r}

roc_plot_muvr <- ggplot()
auc_muvr <- c()

for (i in 1:5){
  
  
  
  # top_var <- muvr_list[[i]]$VIP %>% rownames()
  # 
  # var <- cbind( G_fil[(samples_metadata_fil[folds[[i]],] %>% pull (SXS)), ],
  #                     betas_fil[(samples_metadata_fil[folds[[i]], ] 
  #                                       %>% pull (id)), ])  %>% colnames
  # 
  # new_var <- setdiff (top_var, var)
  # 
  # print (new_var)
  # 
  # G_fil <- cbind (G_fil, G[rownames(G_fil),new_var] )

  
  ### MUVR ROC PLOT ###
  print (i)

  # Definition again of the test variables for the models. 
  y_test_cv <- samples_metadata_fil[folds[[i]],]$target
  
  x_test_cv <- cbind( G[(samples_metadata_fil[folds[[i]],] %>% pull (SXS)), ],
                    betas_fil[(samples_metadata_fil[folds[[i]], ] 
                                      %>% pull (id)), ])[, muvr_list[[i]]$VIP %>% rownames()]
  
  
  prediction_model_cv <- predMV(muvr_list[[i]], x_test_cv, model = "mid")
  
  roc_data <-  roc(y_test_cv, prediction_model_cv$yPred[,2])
  
  roc_plot_muvr <- roc_plot_muvr + geom_path(data = data.frame(x = 1 - roc_data$specificities,
                                y = roc_data$sensitivities,
                                Fold = sprintf("%s, auc = %s", i, round(roc_data$auc,3))),
              aes(x = x, y = y, col = Fold), alpha = 0.8, size = 1.1)  


  cm <- caret::confusionMatrix(as.factor(prediction_model_cv$yClass), as.factor(y_test_cv))
  cm
  auc_muvr <- c(auc_muvr, roc_data$auc[[1]])



  print (i)
}

mean(auc_muvr)#
sd(auc_muvr)

```

mean auc: 0.66
sd auc: 0.0074



```{r}

roc_plot_muvr + labs(x = "1 - Specificity", y = "Sensitivity", title = "ROC Curve") +
    geom_abline(intercept = 0, slope = 1, linetype = "dashed", alpha = 0.5) +
    scale_x_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    theme_bw(base_size = 15)

ggsave("muvr_Methy_SNPs_rocplot.png", path = path_output, units = "px", width = 2187, height = 1350)

```




### 1.2 Methylation + PRS

#### -Preprocessing

Subsample the metadata df to remove only samples that don't have PRS calculated. 
```{r}
samples_metadata_fil_prs <- samples_metadata_fil %>% filter (SXS %in% rownames (mt_all_prs))
```


Selection of the methylation variables with highest variance
```{r}
beta_variance <- apply (beta_values, 1, var)

#Select only values with high variance (top 50%)
betas_fil <- beta_values[ beta_variance >  quantile (beta_variance, 0.50),]

betas_fil <- t(betas_fil[,samples_metadata_fil_prs$id])

```

Selection of PRS samples
```{r}
mt_prs <- mt_all_prs[samples_metadata_fil_prs$SXS, ]

```



#### -Model

CV of the RF model. Featre selection is included in the model.
```{r}
#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil_prs$target, k = 5)

muvr_list <- list()
rf_list <- list()
x_test_topvar <- list()

t1 <- Sys.time()
for (i in 1:5) {
  
  y_train_cv <- samples_metadata_fil_prs[-folds[[i]],]$target
  y_test_cv <- samples_metadata_fil_prs[folds[[i]],]$target
  

  ##PRS data: 
  prs_cv_train <- mt_prs[-folds[[i]],]
  prs_cv_test <- mt_prs[folds[[i]],]
  
  ##Methylation data: 
  
  #Transpose de matrix and select train and test samples
  betas_fil_cv_train <- betas_fil[(samples_metadata_fil_prs[-folds[[i]],] 
                                       %>% pull (id)),]
  betas_fil_cv_test <- betas_fil[(samples_metadata_fil_prs[folds[[i]], ] 
                                      %>% pull (id)), ]
  
  # Select top variables with lower pval
  pval_betas <- apply (betas_fil_cv_train,
                       2,
                       function (j) { t.test (j ~ y_train_cv)$p.val})
  betas_fil_cv_train <- betas_fil_cv_train [, sort (pval_betas, index.return = TRUE)$ix[1:4000]]
  betas_fil_cv_test <- betas_fil_cv_test [, sort (pval_betas, index.return = TRUE)$ix[1:4000]]
  
  
  
  
  print (dim (prs_cv_train))
  print (dim (betas_fil_cv_train))
  ###MUVR
  
  cl = makeCluster(10)
  doParallel::registerDoParallel(cl)
  
  model_muvr_cv <- MUVR( X = cbind (prs_cv_train, betas_fil_cv_train), 
                         Y = y_train_cv, 
                         nRep = 10, 
                         nOuter = 5, 
                         varRatio = 0.7, 
                         method = "RF", 
                         fitness = "AUROC", 
                         modReturn = TRUE)
  
  stopCluster(cl)
  
  #Selecting the most important features according to the
  top_var_cv <- getVIP(model_muvr_cv, model = "mid")$name
  
  
  # ###RF caret: 
  # 
  # 
  # cl <- makeCluster(5)
  # doParallel::registerDoParallel(cl)
  # 
  # tuneGrid <- expand.grid(.mtry = c(2, 8, 20, 50))
  # train_control <- trainControl(method = "cv",
  #                               number = 5,
  #                               verboseIter = TRUE,
  #                               returnResamp = "final",
  #                               classProbs = TRUE,
  #                               summaryFunction = twoClassSummary
  # )
  # rf_fit_cv <- caret::train(x = cbind(G_fil_cv_train, betas_fil_cv_train)[,top_var_cv],
  #                           make.names(y_train_cv),
  #                           method = "rf",
  #                           trControl = train_control, 
  #                           tuneGrid = tuneGrid, 
  #                           ntree = 200
  # )
  # stopCluster(cl)
  
  

  #return models
  muvr_list[[i]] <-  model_muvr_cv
  # rf_list[[i]] <- rf_fit_cv
  x_test_topvar[[i]] <- cbind(prs_cv_test, betas_fil_cv_test)[,top_var_cv]
  
  #Show time taken
  print (Sys.time() - t1)
}

```


Save results
```{r}
saveRDS(list (muvr_list, x_test_topvar), file = paste(path_output, "rds","MUVR_Methy_PRS.rds", sep = "/"))
```

Number of common variables: 

```{r}
top_variables_list <- lapply (x_test_topvar, colnames)
common_top_variables <- Reduce (intersect, top_variables_list )

print (common_top_variables)
#intersect (top_variables_list[[1]], top_variables_list[[2]])
#intersect (intersect (top_variables_list[[1]], top_variables_list[[2]]), top_variables_list[[3]])


sapply (top_variables_list, length) 


#Variables that appear in 3 of the folds
common_variables <- unique (unlist (top_variables_list))
common_top_variables <-  common_variables[sapply 
                                          ( common_variables,
                                            function(x){
                                              sum (sapply (top_variables_list, function (y) x %in% y))}) >= 4]

length (common_top_variables)
common_top_variables
# top_variables_list

```



#### -Plots

```{r}

#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil_prs$target, k = 5)

roc_plot_muvr <- ggplot()
auc_muvr <- c()

for (i in 1:5){
  
  ### MUVR ROC PLOT ###
  
  

  # Definition again of the test variables for the models. 
  y_test_cv <- samples_metadata_fil_prs[folds[[i]],]$target
  
  x_test_cv <- cbind( mt_prs[folds[[i]],],
                      betas_fil[folds[[i]], ])[, muvr_list[[i]]$VIP %>% rownames()]
  
  
  prediction_model_cv <- predMV(muvr_list[[i]], x_test_cv, model = "mid")
  
  roc_data <-  roc(y_test_cv, prediction_model_cv$yPred[,2])
  
  roc_plot_muvr <- roc_plot_muvr + geom_path(data = data.frame(x = 1 - roc_data$specificities,
                                y = roc_data$sensitivities,
                                Fold = sprintf("%s, auc = %s", i, round(roc_data$auc,3))),
              aes(x = x, y = y, col = Fold), alpha = 0.8, size = 1.1)  


  cm <- caret::confusionMatrix(as.factor(prediction_model_cv$yClass), as.factor(y_test_cv))
  cm
  auc_muvr <- c(auc_muvr, roc_data$auc[[1]])

  print (i)
  
}
```


```{r}
roc_plot_muvr + labs(x = "1 - Specificity", y = "Sensitivity", title = "ROC Curve") +
    geom_abline(intercept = 0, slope = 1, linetype = "dashed", alpha = 0.5) +
    scale_x_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    theme_bw(base_size = 15)


mean (auc_muvr)#
sd (auc_muvr) #

ggsave("muvr_methy_PRS_rocplot.png", path = path_output, units = "px", width = 2187, height = 1350)
```
mean auc: 0.6196
sd auc: 0.0813







### 1.3 Metylation

#### -Preprocessing

Selection of the methylation variables with highest variance

```{r}
beta_variance <- apply (beta_values, 1, var)

#Select only values with high variance (top 50%)
betas_fil <- beta_values[ beta_variance >  quantile (beta_variance, 0.50),]

betas_fil <- t(betas_fil[,samples_metadata_fil$id])

```


#### -Model

CV of the RF model. Feature selection is included in the model.

```{r}
#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = 5)

muvr_list <- list()
rf_list <- list()
x_test_topvar <- list()

t1 <- Sys.time()
for (i in 1:5) {
  
  y_train_cv <- samples_metadata_fil[-folds[[i]],]$target
  y_test_cv <- samples_metadata_fil[folds[[i]],]$target
  
  
  
  ##Methylation data: 
  
  #Transpose de matrix and select train and test samples
  betas_fil_cv_train <- betas_fil[(samples_metadata_fil[-folds[[i]],] 
                                       %>% pull (id)),]
  betas_fil_cv_test <- betas_fil[(samples_metadata_fil[folds[[i]], ] 
                                      %>% pull (id)), ]
  
  # Select top variables with lower pval
  pval_betas <- apply (betas_fil_cv_train,
                       2,
                       function (j) { t.test (j ~ y_train_cv)$p.val})
  betas_fil_cv_train <- betas_fil_cv_train [, sort (pval_betas, index.return = TRUE)$ix[1:4000]]
  betas_fil_cv_test <- betas_fil_cv_test [, sort (pval_betas, index.return = TRUE)$ix[1:4000]]
  
  
  ###MUVR
  
  cl = makeCluster(10)
  doParallel::registerDoParallel(cl)
  
  model_muvr_cv <- MUVR( X =  betas_fil_cv_train, 
                         Y = y_train_cv, 
                         nRep = 10, 
                         nOuter = 5, 
                         varRatio = 0.7, 
                         method = "RF", 
                         fitness = "AUROC", 
                         modReturn = TRUE)
  
  stopCluster(cl)
  
  #Selecting the most important features according to the
  top_var_cv <- getVIP(model_muvr_cv, model = "mid")$name
  
  
  # ###RF caret: 
  # 
  # 
  # cl <- makeCluster(5)
  # doParallel::registerDoParallel(cl)
  # 
  # tuneGrid <- expand.grid(.mtry = c(2, 8, 20, 50))
  # train_control <- trainControl(method = "cv",
  #                               number = 5,
  #                               verboseIter = TRUE,
  #                               returnResamp = "final",
  #                               classProbs = TRUE,
  #                               summaryFunction = twoClassSummary
  # )
  # rf_fit_cv <- caret::train(x = cbind(G_fil_cv_train, betas_fil_cv_train)[,top_var_cv],
  #                           make.names(y_train_cv),
  #                           method = "rf",
  #                           trControl = train_control, 
  #                           tuneGrid = tuneGrid, 
  #                           ntree = 200
  # )
  # stopCluster(cl)
  
  
  #return models
  muvr_list[[i]] <-  model_muvr_cv
  # rf_list[[i]] <- rf_fit_cv
  x_test_topvar[[i]] <-  betas_fil_cv_test[,top_var_cv]
  
  #Show time taken
  Sys.time() - t1
}

```



Save results
```{r}
saveRDS(list (muvr_list, x_test_topvar), file = paste(path_output, "rds","MUVR_Methy.rds", sep = "/"))
```


load results
```{r}
list_results <- readRDS (paste(path_output, "rds","MUVR_Methy.rds", sep = "/"))

muvr_list <- list_results[[1]]
x_test_topvar <- list_results[[2]]

#Fold generation again 
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = n_folds)
```


Number of common variables: 
```{r}
top_variables_list <- lapply (x_test_topvar, colnames)
common_top_variables <- Reduce (intersect, top_variables_list )


#intersect (top_variables_list[[1]], top_variables_list[[2]])
#intersect (intersect (top_variables_list[[1]], top_variables_list[[2]]), top_variables_list[[3]])


sapply (top_variables_list, length) 
print (common_top_variables)


#Variables that appear in 3 of the folds
common_variables <- unique (unlist (top_variables_list))
common_top_variables <-  common_variables[sapply 
                                          ( common_variables,
                                            function(x){
                                              sum (sapply (top_variables_list, function (y) x %in% y))}) >= 4]

length (common_top_variables)
common_top_variables
# top_variables_list

```
3 variables selected in all the folds ("cg08582801" "cg14120049" "cg11144229")
9 variables selected in 4 different folds ("cg08582801" "cg25669494" "cg14120049" "cg15815084" "cg11144229" "cg12634080" "cg21020855" "cg03736826" "cg12002139")



#### -Plots

```{r}

#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = 5)

roc_plot_muvr <- ggplot()
auc_muvr <- c()

for (i in 1:5){
  
  ### MUVR ROC PLOT ###

  # Definition again of the test variables for the models. 
  y_test_cv <- samples_metadata_fil[folds[[i]],]$target
  x_test_cv <- betas_fil[folds[[i]], ][, muvr_list[[i]]$VIP %>% rownames()]
  
  
  prediction_model_cv <- predMV(muvr_list[[i]], x_test_cv, model = "mid")
  
  roc_data <-  roc(y_test_cv, prediction_model_cv$yPred[,2])
  
  roc_plot_muvr <- roc_plot_muvr + geom_path(data = data.frame(x = 1 - roc_data$specificities,
                                y = roc_data$sensitivities,
                                Fold = sprintf("%s, auc = %s", i, round(roc_data$auc,3))),
              aes(x = x, y = y, col = Fold), alpha = 0.8, size = 1.1)  


  cm <- caret::confusionMatrix(as.factor(prediction_model_cv$yClass), as.factor(y_test_cv))
  cm
  auc_muvr <- c(auc_muvr, roc_data$auc[[1]])



  print (i)
  
}
```


```{r}
roc_plot_muvr + labs(x = "1 - Specificity", y = "Sensitivity", title = "ROC Curve") +
    geom_abline(intercept = 0, slope = 1, linetype = "dashed", alpha = 0.5) +
    scale_x_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    theme_bw(base_size = 15)


mean (auc_muvr) #
sd (auc_muvr)  #

ggsave("muvr_Methy_rocplot.png", path = path_output, units = "px", width = 2187, height = 1350)
```

mean auc: 0.6575
sd auc= 0.078



### 1.3.2 Methlation limma pval filtering

#### -Preprocessing

Selection of the methylation variables with highest variance

```{r}
beta_variance <- apply (beta_values, 1, var)

#Select only values with high variance (top 50%)
betas_fil <- beta_values[ beta_variance >  quantile (beta_variance, 0.50),]

betas_fil <- t(betas_fil[,samples_metadata_fil$id])#transposed
mval_fil <- mval[colnames (betas_fil),samples_metadata_fil$id]

```


#### -Model

CV of the RF model. Feature selection is included in the model.

```{r}
#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = 5)


## Add cell type distribution to data
cell_estimate <- readRDS("OUTPUT/methylation_preprocessing/rdsDir/cell_estimate.rds")
samples_metadata_fil <- cell_estimate %>% as.data.frame() %>% rownames_to_column("id") %>% left_join(samples_metadata_fil, .)


muvr_list <- list()
rf_list <- list()
x_test_topvar <- list()

t1 <- Sys.time()
for (i in 1:5) {
  
  y_train_cv <- samples_metadata_fil[-folds[[i]],]$target
  y_test_cv <- samples_metadata_fil[folds[[i]],]$target
  
  
  
  ##Methylation data: 
  
  #Transpose de matrix and select train and test samples
  betas_fil_cv_train <- betas_fil[(samples_metadata_fil[-folds[[i]],] 
                                       %>% pull (id)),]
  betas_fil_cv_test <- betas_fil[(samples_metadata_fil[folds[[i]], ] 
                                      %>% pull (id)), ]
  
  
  modelmat <- model.matrix (~ 0 + target + Sex + Age + Array + CD8T + CD4T + NK + Bcell + Mono + Gran,
                            data = samples_metadata_fil[-folds[[i]],] %>% 
                              mutate (Sex = ifelse (Sex == 1, "Male", "Female"), 
                                      target = make.names (target)
                                      )
                            )
  fit <- lmFit (mval_fil[, -folds[[i]]], 
                modelmat)
  
  contrasts <- makeContrasts (targetExposed- targetNon.exposed, levels = modelmat)

  fit <- contrasts.fit (fit, contrasts )
  
  fit <- eBayes(fit)

  top_table <- topTable(fit, number = "inf", sort.by = "p", adjust.method = "BH")
  
  
  
  betas_fil_cv_train <- betas_fil_cv_train [,  rownames(top_table)[1:4000]]
  betas_fil_cv_test <- betas_fil_cv_test [,  rownames(top_table)[1:4000]]
  
  
  # Select top variables with lower pval
  pval_betas <- apply (betas_fil_cv_train,
                       2,
                       function (j) { t.test (j ~ y_train_cv)$p.val})
  betas_fil_cv_train <- betas_fil_cv_train [, sort (pval_betas, index.return = TRUE)$ix[1:4000]]
  betas_fil_cv_test <- betas_fil_cv_test [, sort (pval_betas, index.return = TRUE)$ix[1:4000]]
  
  
  ###MUVR
  
  cl = makeCluster(10)
  doParallel::registerDoParallel(cl)
  
  model_muvr_cv <- MUVR( X =  betas_fil_cv_train, 
                         Y = y_train_cv, 
                         nRep = 10, 
                         nOuter = 5, 
                         varRatio = 0.7, 
                         method = "RF", 
                         fitness = "AUROC", 
                         modReturn = TRUE)
  
  stopCluster(cl)
  
  #Selecting the most important features according to the
  top_var_cv <- getVIP(model_muvr_cv, model = "mid")$name
  
  
  
  #return models
  muvr_list[[i]] <-  model_muvr_cv
  # rf_list[[i]] <- rf_fit_cv
  x_test_topvar[[i]] <-  betas_fil_cv_test[,top_var_cv]
  
  #Show time taken
  Sys.time() - t1
}

```



Save results
```{r}
saveRDS(list (muvr_list, x_test_topvar), file = paste(path_output, "rds","MUVR_Methy_limma_filter.rds", sep = "/"))
```


load results
```{r}
list_results <- readRDS ( paste(path_output, "rds","MUVR_Methy_limma_filter.rds", sep = "/"))

muvr_list <- list_results[[1]]
x_test_topvar <- list_results[[2]]

#Fold generation again 
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = n_folds)
```



Number of common variables: 
```{r}
top_variables_list <- lapply (x_test_topvar, colnames)
common_top_variables <- Reduce (intersect, top_variables_list )


#intersect (top_variables_list[[1]], top_variables_list[[2]])
#intersect (intersect (top_variables_list[[1]], top_variables_list[[2]]), top_variables_list[[3]])


sapply (top_variables_list, length) 
print (common_top_variables)


#Variables that appear in 3 of the folds
common_variables <- unique (unlist (top_variables_list))
common_top_variables <-  common_variables[sapply 
                                          ( common_variables,
                                            function(x){
                                              sum (sapply (top_variables_list, function (y) x %in% y))}) >= 4]

length (common_top_variables)
common_top_variables
# top_variables_list

```




#### -Plots

```{r}

#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = 5)

roc_plot_muvr <- ggplot()
auc_muvr <- c()

for (i in 1:5){
  
  ### MUVR ROC PLOT ###

  # Definition again of the test variables for the models. 
  y_test_cv <- samples_metadata_fil[folds[[i]],]$target
  x_test_cv <- betas_fil[folds[[i]], ][, muvr_list[[i]]$VIP %>% rownames()]
  
  
  prediction_model_cv <- predMV(muvr_list[[i]], x_test_cv, model = "mid")
  
  roc_data <-  roc(y_test_cv, prediction_model_cv$yPred[,2])
  
  roc_plot_muvr <- roc_plot_muvr + geom_path(data = data.frame(x = 1 - roc_data$specificities,
                                y = roc_data$sensitivities,
                                Fold = sprintf("%s, auc = %s", i, round(roc_data$auc,3))),
              aes(x = x, y = y, col = Fold), alpha = 0.8, size = 1.1)  


  cm <- caret::confusionMatrix(as.factor(prediction_model_cv$yClass), as.factor(y_test_cv))
  cm
  auc_muvr <- c(auc_muvr, roc_data$auc[[1]])



  print (i)
  
}
```


```{r}
roc_plot_muvr + labs(x = "1 - Specificity", y = "Sensitivity", title = "ROC Curve") +
    geom_abline(intercept = 0, slope = 1, linetype = "dashed", alpha = 0.5) +
    scale_x_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    theme_bw(base_size = 15)


mean (auc_muvr) 
sd (auc_muvr)  

ggsave("muvr_Methy_limma_filter_rocplot.png", path = path_output, units = "px", width = 2187, height = 1350)
```
mean auc: 0.666756
sd auc: 0.04715202



### 1.4 Genomic


#### -Preprocessing


Selection of SNPs with high maf

```{r}

G_fil <- G[samples_metadata_fil$SXS, which (maf > 0.1)]

```

#### -Model

CV of the RF model. Featre selection is included in the model.

```{r}
#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = 5)

muvr_list <- list()
rf_list <- list()
x_test_topvar <- list()

t1 <- Sys.time()
for (i in 1:5) {
  
  y_train_cv <- samples_metadata_fil[-folds[[i]],]$target
  y_test_cv <- samples_metadata_fil[folds[[i]],]$target
  
  
  
  ##Genomic data: 
  #Based on previous G_fil2 and betas_fil2, that were filtered based on variance or maf
  
  G_fil_cv_train <- G_fil[ (samples_metadata_fil[-folds[[i]],] %>% pull (SXS)),]
  G_fil_cv_test <- G_fil[(samples_metadata_fil[folds[[i]],] %>% pull (SXS)), ]
  
  x_square <- function (j){
    chi <- chisq.test(j, y_train_cv)
    return (chi$p.val)
  }
  pval_xsqr <- apply (G_fil_cv_train,2, function(j){chisq.test(j, y_train_cv)$p.val})
  #Selection of the top 5000 snps that correlate with target. 
  G_fil_cv_train <-  G_fil_cv_train[, sort(pval_xsqr, index.return = TRUE)$ix[1:4000]]
  G_fil_cv_test <-  G_fil_cv_test[, sort(pval_xsqr, index.return = TRUE)$ix[1:4000]]
  

  
  
  
  
  
  ###MUVR
  
  cl = makeCluster(10)
  doParallel::registerDoParallel(cl)
  
  model_muvr_cv <- MUVR( X = G_fil_cv_train, 
                         Y = y_train_cv, 
                         nRep = 10, 
                         nOuter = 5, 
                         varRatio = 0.7, 
                         method = "RF", 
                         fitness = "AUROC", 
                         modReturn = TRUE)
  
  stopCluster(cl)
  
  #Selecting the most important features according to the
  top_var_cv <- getVIP(model_muvr_cv, model = "mid")$name
  
  
  # ###RF caret: 
  # 
  # 
  # cl <- makeCluster(5)
  # doParallel::registerDoParallel(cl)
  # 
  # tuneGrid <- expand.grid(.mtry = c(2, 8, 20, 50))
  # train_control <- trainControl(method = "cv",
  #                               number = 5,
  #                               verboseIter = TRUE,
  #                               returnResamp = "final",
  #                               classProbs = TRUE,
  #                               summaryFunction = twoClassSummary
  # )
  # rf_fit_cv <- caret::train(x = cbind(G_fil_cv_train, betas_fil_cv_train)[,top_var_cv],
  #                           make.names(y_train_cv),
  #                           method = "rf",
  #                           trControl = train_control, 
  #                           tuneGrid = tuneGrid, 
  #                           ntree = 200
  # )
  # stopCluster(cl)
  
  

  #return models
  muvr_list[[i]] <-  model_muvr_cv
  # rf_list[[i]] <- rf_fit_cv
  x_test_topvar[[i]] <- G_fil_cv_test[,top_var_cv]
  
  #Show time taken
  Sys.time() - t1
}

```

Save results

```{r}
saveRDS(list (muvr_list, x_test_topvar), file = paste(path_output, "rds","MUVR_SNPs.rds", sep = "/"))
```


Load model: 

```{r}
list_results <- readRDS(file = paste(path_output, "rds","MUVR_SNPs.rds", sep = "/"))

muvr_list <- list_results[[1]]
x_test_topvar <- list_results[[2]]
```





Top variables

```{r}
top_variables_list <- lapply (x_test_topvar, colnames)

common_top_variables <- Reduce (intersect, top_variables_list )
print (common_top_variables)

sapply (top_variables_list, length) 

# #Percentage of genomic variables in top variables
# sapply (top_variables_list, function(x){round (length (x[grep("rs", x)]) * 100/length (x), 2)}) %>% 
# 
# #Number of genomic variables in top variables
# sapply (top_variables_list, function(x){length (x[grep("rs", x)]) }) 


#Variables that appear in 3 of the folds
common_variables <- unique (unlist (top_variables_list))
common_top_variables <-  common_variables[sapply 
                                          ( common_variables,
                                            function(x){
                                              sum (sapply (top_variables_list, function (y) x %in% y))}) >= 4]

print (common_top_variables)

# print (top_variables_list)

```
Many features selected in each model (465). 19 features selected in all folds and 64 selected in 4 of them. 


#### -Plots 
Roc plots of the MUVR models
```{r}
#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil$target, k = 5)

roc_plot_muvr <- ggplot()
auc_muvr <- c()

for (i in 1:5){
  
  ### MUVR ROC PLOT ###

  # Definition again of the test variables for the models. 
  y_test_cv <- samples_metadata_fil[folds[[i]],]$target
  
  x_test_cv <- G_fil[(samples_metadata_fil[folds[[i]],] %>% pull (SXS)), ][, muvr_list[[i]]$VIP %>% rownames()]
  
  
  prediction_model_cv <- predMV(muvr_list[[i]], x_test_cv, model = "mid")
  
  roc_data <-  roc(y_test_cv, prediction_model_cv$yPred[,2])
  
  roc_plot_muvr <- roc_plot_muvr + geom_path(data = data.frame(x = 1 - roc_data$specificities,
                                y = roc_data$sensitivities,
                                Fold = sprintf("%s, auc = %s", i, round(roc_data$auc,3))),
              aes(x = x, y = y, col = Fold), alpha = 0.8, size = 1.1)  


  cm <- caret::confusionMatrix(as.factor(prediction_model_cv$yClass), as.factor(y_test_cv))
  cm
  auc_muvr <- c(auc_muvr, roc_data$auc[[1]])



  print (i)
}

mean(auc_muvr)#
sd(auc_muvr)#

```
0.586
0.067




```{r}

roc_plot_muvr + labs(x = "1 - Specificity", y = "Sensitivity", title = "ROC Curve") +
    geom_abline(intercept = 0, slope = 1, linetype = "dashed", alpha = 0.5) +
    scale_x_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    theme_bw(base_size = 15)

ggsave("muvr_SNPs_rocplot.png", path = path_output, units = "px", width = 2187, height = 1350)

```



### 1.5 PRS

#### -Preprocessing


Subsample the metadata df to remove only samples that don't have PRS calculated. 
```{r}
samples_metadata_fil_prs <- samples_metadata_fil %>% filter (SXS %in% rownames (mt_all_prs))
```

Selection of PRS samples
```{r}
mt_prs <- mt_all_prs[samples_metadata_fil_prs$SXS, ]
```



#### -Model

CV of the RF model. Feature selection is included in the model.

```{r}
#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil_prs$target, k = 5)

muvr_list <- list()
rf_list <- list()
x_test_topvar <- list()

t1 <- Sys.time()
for (i in 1:5) {
  
  y_train_cv <- samples_metadata_fil_prs[-folds[[i]],]$target
  y_test_cv <- samples_metadata_fil_prs[folds[[i]],]$target
  
  
  

  ##PRS data: 
  prs_cv_train <- mt_prs[-folds[[i]],]
  prs_cv_test <- mt_prs[folds[[i]],]

  
  
  
  
  ###MUVR
  
  cl = makeCluster(10)
  doParallel::registerDoParallel(cl)
  
  model_muvr_cv <- MUVR( X = prs_cv_train, 
                         Y = y_train_cv, 
                         nRep = 10, 
                         nOuter = 5, 
                         varRatio = 0.7, 
                         method = "RF", 
                         fitness = "AUROC", 
                         modReturn = TRUE)
  
  stopCluster(cl)
  
  #Selecting the most important features according to the
  top_var_cv <- getVIP(model_muvr_cv, model = "mid")$name
  
  
  # ###RF caret: 
  # 
  # 
  # cl <- makeCluster(5)
  # doParallel::registerDoParallel(cl)
  # 
  # tuneGrid <- expand.grid(.mtry = c(2, 8, 20, 50))
  # train_control <- trainControl(method = "cv",
  #                               number = 5,
  #                               verboseIter = TRUE,
  #                               returnResamp = "final",
  #                               classProbs = TRUE,
  #                               summaryFunction = twoClassSummary
  # )
  # rf_fit_cv <- caret::train(x = cbind(G_fil_cv_train, betas_fil_cv_train)[,top_var_cv],
  #                           make.names(y_train_cv),
  #                           method = "rf",
  #                           trControl = train_control, 
  #                           tuneGrid = tuneGrid, 
  #                           ntree = 200
  # )
  # stopCluster(cl)
  
  

  #return models
  muvr_list[[i]] <-  model_muvr_cv
  # rf_list[[i]] <- rf_fit_cv
  x_test_topvar[[i]] <- prs_cv_test[,top_var_cv]
  
  #Show time taken
  Sys.time() - t1
}

```


Save results
```{r}
saveRDS(list (muvr_list, x_test_topvar), file = paste(path_output, "rds","MUVR_PRS.rds", sep = "/"))
```



Number of common variables: 

```{r}
top_variables_list <- lapply (x_test_topvar, colnames)
common_top_variables <- Reduce (intersect, top_variables_list )


#intersect (top_variables_list[[1]], top_variables_list[[2]])
#intersect (intersect (top_variables_list[[1]], top_variables_list[[2]]), top_variables_list[[3]])


sapply (top_variables_list, length) 
common_top_variables

#Variables that appear in 3 of the folds
common_variables <- unique (unlist (top_variables_list))
common_top_variables <-  common_variables[sapply 
                                          ( common_variables,
                                            function(x){
                                              sum (sapply (top_variables_list, function (y) x %in% y))}) >= 4]

length (common_top_variables) 
mean(sapply (top_variables_list, length))#21.6
common_top_variables
# top_variables_list

```
"Pt_0.0001-Attention_deficit_hyperactivity_disorder" selected in all folds. 
"Pt_0.0001-HOMA-B" and "Pt_0.0001-Non-cancer_illness_code_self-reported_depression" in 4 of the folds



#### -Plots

```{r}

#set seed to have consistent results
set.seed(5)
folds <- caret::createFolds (samples_metadata_fil_prs$target, k = 5)

roc_plot_muvr <- ggplot()
auc_muvr <- c()

for (i in 1:5){
  
  ### MUVR ROC PLOT ###

  # Definition again of the test variables for the models. 
  y_test_cv <- samples_metadata_fil_prs[folds[[i]],]$target
  x_test_cv <- mt_prs[folds[[i]],][, muvr_list[[i]]$VIP %>% rownames()]
  
  
  prediction_model_cv <- predMV(muvr_list[[i]], x_test_cv, model = "mid")
  
  roc_data <-  roc(y_test_cv, prediction_model_cv$yPred[,2])
  
  roc_plot_muvr <- roc_plot_muvr + geom_path(data = data.frame(x = 1 - roc_data$specificities,
                                y = roc_data$sensitivities,
                                Fold = sprintf("%s, auc = %s", i, round(roc_data$auc,3))),
              aes(x = x, y = y, col = Fold), alpha = 0.8, size = 1.1)  


  cm <- caret::confusionMatrix(as.factor(prediction_model_cv$yClass), as.factor(y_test_cv))
  cm
  auc_muvr <- c(auc_muvr, roc_data$auc[[1]])
  
  print (i)
  print (cm)
  
  
}
```


```{r}
roc_plot_muvr + labs(x = "1 - Specificity", y = "Sensitivity", title = "ROC Curve") +
    geom_abline(intercept = 0, slope = 1, linetype = "dashed", alpha = 0.5) +
    scale_x_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    scale_y_continuous(limits = c(0,1), expand = c(0, 0.01)) +
    theme_bw(base_size = 15)


mean (auc_muvr)#0
sd (auc_muvr) #0

ggsave("muvr_PRS_rocplot.png", path = path_output, units = "px", width = 2187, height = 1350)
```
mean auc = 0.5486
sd auc = 0.07121



```{r}
sessionInfo()
```


