---
title: "R Notebook"
output: html_notebook
---


Differential methylation analysis of the top features selected by the RF model (usin only methylation data) of 
Mid Exposed patients vs Post Famine patients

## 0. Set-up

```{r}
library ("tidyverse") 
library ("ggplot2")
library ("ggpubr")
library ("minfi")
library ("MUVR")
library ("doParallel")
library ("bigsnpr")
library ("bigstatsr")
library ("pROC")
library ("randomForest")
library ("caret")
library ("Rtsne")
library ("janitor")
library ("limma")
library ("DMRcate")
library ("Cairo")
library ("ggrepel")

# theme_set(theme_bw())
```


Import data:

```{r}

#Metadata: ----
samples_metadata <- readRDS("OUTPUT/feature_selection/rds/samples_metadata.rds")


#Methylation data: ----
grset <- readRDS("OUTPUT/methylation_preprocessing/rdsDir/grset_normalized_noSNPs_noSex.Rds")
beta_values <- getBeta (grset)
mval <- getM(grset)


#Model to load
muvr_model <- readRDS ("OUTPUT/top_features_models/rds/muvr_MEvPF.rds")
top_var <- getVIP(muvr_model, model = "max")$name


#Model limma: 
muvr_model_limma <- readRDS("OUTPUT/top_features_models/rds/muvr_MEvPF_limma_filter.rds")
top_var_limma <- getVIP(muvr_model_limma, model = "max")$name

```



```{r}

samples_metadata_fil <- samples_metadata %>% filter (id %in% colnames (beta_values))

betas_fil <- beta_values[top_var, samples_metadata_fil$id]
mval_fil <- mval[top_var, samples_metadata_fil$id]


#Output folder: 

out_dir <- "OUTPUT/top_methylation_features/MEvPF"

# Check if output folder exist
if (!dir.exists(out_dir)) {
  dir.create(out_dir)
} 
```




Data frame with annotation of the top probes: 
```{r}

anno_gr <- makeGRangesFromDataFrame(df = minfi::getAnnotation(grset), 
                                    keep.extra.columns = T, 
                                    seqnames.field = "chr", 
                                    start.field = "pos", 
                                    end.field = "pos")
## dmps (of top variables)
# comp_top <- cbind(top_table[top_var,],
#                   Beta = coef(fit2)[top_var, 2],
#                   data.frame(anno_gr[top_var, ]))

comp_top <- data.frame (anno_gr[top_var,])

anno_df <- as.data.frame (anno_gr)
```




## 1. DM analysis

Linear models with mvalues


With covariates: 
```{r}

cell_estimate <- readRDS("OUTPUT/methylation_preprocessing/rdsDir/cell_estimate.rds")

samples_metadata_fil <- cell_estimate %>% as.data.frame() %>% rownames_to_column("id") %>% left_join(samples_metadata_fil, .)

modelmat <- model.matrix(~ 0 + target + Sex + Age + Array + CD8T + CD4T + NK + Bcell + Mono + Gran,
                         data = samples_metadata_fil %>%
                           mutate (Sex = ifelse (Sex == 1, "Male", "Female"),
                                   target = make.names(target)) %>% 
                           filter (Timeframe_non == "Postfamine" | Exposure =="Mid"))


colnames(modelmat)[1:4] <- c("Exposed", "Non.Exposed", "Male", "Age")

fit.2 <- lmFit(mval[, samples_metadata_fil %>% filter (Timeframe_non == "Postfamine" | Exposure == "Mid") %>% pull(id)], 
               modelmat)

contrasts <- makeContrasts (EvNE = Exposed - Non.Exposed,
                            levels = modelmat)

fit2 <- contrasts.fit (fit.2, contrasts )
fit2 <- eBayes(fit2)



top_table_celltypes <- topTable(fit2, number = "inf", adjust.method = "fdr", sort.by = "p")
# print (top_table)



#Volcano plot
vp1 <- ggplot (top_table_celltypes, aes (y = -log10(P.Value), x = logFC)) + 
  geom_point(size = 0.5) +   
  geom_point (data = top_table_celltypes[top_var_limma,], aes (x = logFC, y = -log10(P.Value)), color = "red")+
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_point( data = . %>% filter (adj.P.Val < 0.05), aes (y = -log10(P.Value), x = logFC), color = "red")+
  labs (title = "ME vs Postfamine") +
  xlab ("logFC (Exposed/Non-Exposed)")
print (vp1)


#Volcano plot
vp1 <- ggplot (top_table_celltypes, aes (y = -log10(P.Value), x = logFC)) + 
  geom_point(size = 0.5) +   
  geom_point (data = top_table_celltypes[top_var,], aes (x = logFC, y = -log10(P.Value)), color = "red")+
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_point( data = . %>% filter (adj.P.Val < 0.05), aes (y = -log10(P.Value), x = logFC), color = "red")+
  labs (title = "ME vs Postfamine") +
  xlab ("logFC (Exposed/Non-Exposed)")
print (vp1)

#Volcano plot adj Pval
vp2 <- ggplot (top_table_celltypes, aes (y = -log10(adj.P.Val), x = logFC)) + 
  geom_point() + 
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_point( data = . %>% filter (adj.P.Val < 0.05), aes (y = -log10(adj.P.Val), x = logFC), color = "red")+
  labs (title = "ME vs Postfamine") +
  xlab ("logFC (Exposed/Non-Exposed)")
print (vp2)




```

Without cell type: 
```{r}

modelmat <- model.matrix(~ 0 + target + Sex + Age + Array,
                         data = samples_metadata_fil %>%
                           mutate (Sex = ifelse (Sex == 1, "Male", "Female"),
                                   target = make.names(target)) %>% 
                           filter (Timeframe_non == "Postfamine" |  Exposure == "Mid"))


colnames(modelmat)[1:2] <- c("Exposed", "Non.Exposed")

#Filtering of the appropiate samples
fit.2 <- lmFit(mval[, samples_metadata_fil %>% filter (Timeframe_non == "Postfamine" |  Exposure == "Mid") %>% pull(id)], 
               modelmat)

contrasts <- makeContrasts (EvNE = Exposed - Non.Exposed,
                            levels = modelmat)

fit2 <- contrasts.fit (fit.2, contrasts )
fit2 <- eBayes(fit2)



top_table <- topTable(fit2, number = "inf", adjust.method = "fdr", sort.by = "p")
print (top_table)



#Volcano plot
vp1 <- ggplot (top_table, aes (y = -log10(P.Value), x = logFC)) + 
  geom_point(size = 0.5) +   
  geom_point (data = top_table[top_var,], aes (x = logFC, y = -log10(P.Value)), color = "red")+
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_point( data = . %>% filter (adj.P.Val < 0.05), aes (y = -log10(P.Value), x = logFC), color = "red")+
  labs (title = "ME vs Postfamine") +
  xlab ("logFC (Exposed/Non-Exposed)")
print (vp1)

ggsave ("volcano_plot.png", path = out_dir, units = "in", height = 5, width = 5)

#Volcano plot adj Pval
vp2 <- ggplot (top_table, aes (y = -log10(adj.P.Val), x = logFC)) + 
  geom_point() + 
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_point( data = . %>% filter (adj.P.Val < 0.05), aes (y = -log10(adj.P.Val), x = logFC), color = "red")+
  labs (title = "ME vs Postfamine") +
  xlab ("logFC (Exposed/Non-Exposed)")
print (vp2)


```

Without covariates: 
```{r}

modelmat <- model.matrix(~ 0 + target,
                         data = samples_metadata_fil %>%
                           mutate (Sex = ifelse (Sex == 1, "Male", "Female"),
                                   target = make.names(target)) %>% 
                           filter (Timeframe_non == "Postfamine" |  Exposure == "Mid"))


colnames(modelmat)[1:2] <- c("Exposed", "Non.Exposed")

#Filtering of the appropiate samples
fit.2 <- lmFit(mval[, samples_metadata_fil %>% filter (Timeframe_non == "Postfamine" |  Exposure == "Mid") %>% pull(id)], 
               modelmat)

contrasts <- makeContrasts (EvNE = Exposed - Non.Exposed,
                            levels = modelmat)

fit2 <- contrasts.fit (fit.2, contrasts )
fit2 <- eBayes(fit2)



top_table_nocov <- topTable(fit2, number = "inf", adjust.method = "fdr", sort.by = "p")
print (top_table)



#Volcano plot
vp1 <- ggplot (top_table_nocov, aes (y = -log10(P.Value), x = logFC)) + 
  geom_point(size = 0.5) +   
  geom_point (data = top_table[top_var,], aes (x = logFC, y = -log10(P.Value)), color = "red")+
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_point( data = . %>% filter (adj.P.Val < 0.05), aes (y = -log10(P.Value), x = logFC), color = "red")+
  labs (title = "ME vs Postfamine") +
  xlab ("logFC (Exposed/Non-Exposed)")
print (vp1)

#Volcano plot adj Pval
vp2 <- ggplot (top_table_nocov, aes (y = -log10(adj.P.Val), x = logFC)) + 
  geom_point() + 
  geom_hline(yintercept = -log10(0.05), linetype = "dashed") + 
  geom_point( data = . %>% filter (adj.P.Val < 0.05), aes (y = -log10(adj.P.Val), x = logFC), color = "red")+
  labs (title = "ME vs Postfamine") +
  xlab ("logFC (Exposed/Non-Exposed)")
print (vp2)


```



Manhathan plot: 
```{r}

anno_df <- as.data.frame (anno_gr)

#Merge with top_table data
anno_df <- anno_df %>% arrange (seqnames, start) %>% cbind (., top_table_celltypes[anno_df$Name,]) %>% rename (chr = seqnames) %>% mutate (start = as.numeric(start))

anno_df <- 
  anno_df %>% 
  group_by(chr) %>% 
  summarise (chr_len = max(start)) %>% 
  mutate (tot = cumsum (chr_len)-chr_len) %>% 
  left_join(anno_df, . ) %>% 
  mutate (start_cum = start + tot) %>% 
  mutate (top_var = ifelse (Name %in% top_var_limma, "Top var", "No"))




# Prepare X axis
axisdf <- anno_df %>% group_by(chr) %>% summarize(center=( max(start_cum) + min(start_cum) ) / 2 ) %>% mutate (CHR = row_number (center))


# Make the  manhattan plot
ggplot(anno_df, aes(x=start_cum, y=-log10(P.Value))) +
    
    # Show all points
    geom_point( aes(color=as.factor(chr)), alpha=0.6, size=1.3, show.legend = FALSE) +
    scale_color_manual(values = rep(c("grey29", "grey"), 22 )) +
    
    # custom X axis:
    scale_x_continuous( label = axisdf$CHR, breaks= axisdf$center ) +
    scale_y_continuous(expand = c(0, 0) ) +     # remove space between plot area and x axis

    # Add highlighted points
    geom_point(data = . %>% filter (top_var != "No"), color = "Red", size=2, show.legend = FALSE) +
    # 
    # # Add label using ggrepel to avoid overlapping
    # geom_label_repel( data=subset(don, is_annotate=="yes"), aes(label=SNP), size=2) +

    # Custom the theme: + 
    labs (x = "Chromosome")+
    theme_classic(base_size = 15) + 
  theme (axis.text.x = element_text( size = 10) )

  
    # theme( 
    #   # legend.position="none",
    #   panel.border = element_blank(),
    #   panel.grid.major.x = element_blank(),
    #   panel.grid.minor.x = element_blank()
    # )

ggsave ("manhattan_MEvPF.png", path = out_dir, height = 5, width = 9, units = "in")

head (anno_df)

```




```{r}
df_betas_fil <- betas_fil %>% t() %>%  as.data.frame() %>% rownames_to_column("CG") %>%  pivot_longer( contains ("R0"), names_to = "id", values_to = "beta") %>% dplyr::inner_join(., samples_metadata_fil %>% filter (Timeframe_non == "Postfamine" | Exposure == "Mid"), by = "id" )


for (cg in top_var[1:10]){
  cg_plot <- df_betas_fil %>% filter (CG == cg) %>%
  ggplot (aes(y = beta , x = target)) + 
  geom_violin(aes(fill = target), alpha = 0.8) + 
  geom_boxplot (alpha = 0.8, width = 0.2, fill = "white", color = "black") +
  stat_compare_means(comparisons = list(c("Exposed", "Non-exposed")), label = "p.signif")+
  ylim(0,1)+
  labs (subtitle = cg, 
        # subtitle = paste0("BH-adjusted p = ",formatC(top_table[cg,"adj.P.Val"], format = "e", digits = 2)),
        x = NULL,
        y = "Beta values")+
  theme(text =  element_text(size=15))
  
  print (cg_plot) 
}

```



Correlation ranking and pval


```{r}

top_var_ranking <- getVIP(muvr_model_limma, model = "max") %>% rownames_to_column(var = "CG")


top_table_celltypes %>%
  rownames_to_column(var = "CG") %>%
  inner_join(., top_var_ranking) %>% 
  ggplot( aes (x = rank, y = -log10(P.Value)))+ 
  geom_point() + 
  geom_smooth(stat = "smooth", method = "lm")+ 
  stat_cor() + 
  theme_bw(base_size = 14) +
  labs (x = "Rank top variables")
  
ggsave ("pval_ranking_regression_MEvPF.png", path = out_dir, units = "in")
  
```

















## 2. Genes methylation plots



```{r}
comp_top <- data.frame (anno_gr[top_var_limma,])
```


```{r}

comp_top_genes <-  lapply(strsplit(comp_top$UCSC_RefGene_Name, ";"), unique)

names(comp_top_genes) <- rownames(comp_top)

comp_top_genes_nz <- comp_top_genes[which(lapply(comp_top_genes, length) != 0)]

comp_top_genes_nz_df <- data.frame(CpG = gsub("(^cg[0-9]{8}).+$", "\\1",
                                              names(unlist(comp_top_genes_nz))), 
                                   Gene = unlist(comp_top_genes_nz))


# comp_top_genes_nz_df$pvalue <- top_table[comp_top_genes_nz_df$CpG, "P.Value"]

comp_top_genes_nz_list <- split(comp_top_genes_nz_df, comp_top_genes_nz_df$Gene)
comp_top_genes_nz_list <- comp_top_genes_nz_list[unlist(lapply(comp_top_genes_nz_list, nrow)) != 1]

# comp_top_fisher <- lapply(X = comp_top_genes_nz_list, FUN = function(gene){
#     unlist(EmpiricalBrownsMethod::empiricalBrownsMethod(data_matrix = beta_values[gene$CpG,],
#                                                         p_values = gene$pvalue,
#                                                         extra_info = T))
#   })


```
max model t.test filtering
115 CpG probes (from the 160 top_var EMEvPF) are associated with 114 genes. 1 genes have associated 2 different CpG probes. 

max model limma filtering:
164 CpG probes (from the 229) are associated with 154 genes. There are 4 genes mapped by 2 different probes and 2 by 3 different probes
4x2, 2x3

Gene plot function: 
```{r gene plot function}

dmg_plot_continuous <- function(grset, beta_values, samples_metadata_fil, gene, top_var) {
  
  require(ggbio)
  require(TxDb.Hsapiens.UCSC.hg19.knownGene)
  
  data(genesymbol, package = "biovizBase")
  
  
  anno_gr <- makeGRangesFromDataFrame(df = minfi::getAnnotation(grset), 
                                    keep.extra.columns = T, 
                                    seqnames.field = "chr", 
                                    start.field = "pos", 
                                    end.field = "pos")

  #annotation of the gene
  gen_anno <- anno_gr[grep(paste0("(^|;)", as.character(gene), "($|;)"),
                        anno_gr$UCSC_RefGene_Name),]


  gen_df <- as.data.frame(gen_anno) %>%
    rownames_to_column ("CG") %>%
    left_join (., beta_values[gen_anno$Name, colnames (betas_fil)] %>%
                 as.data.frame() %>%
                 rownames_to_column("CG")) %>%
    pivot_longer( contains ("R0"), names_to = "id", values_to = "beta") %>%
    left_join(., samples_metadata_fil)

  
  
  # Beta plot
  data_beta <- gen_df %>%
    group_by(CG, target) %>%
    summarise (mean_beta = mean (beta), start = mean (start)) %>%
    as.data.frame() %>% 
    split (., .$target)
  
  data_beta_E <- data_beta$Exposed
  data_beta_NE <- data_beta$`Non-exposed`
  
  plot_beta <- ggplot () + 
    geom_line ( data = data_beta_E, aes (x = start, y = mean_beta, color = target)) +
    geom_line ( data = data_beta_NE, aes (x = start, y = mean_beta, color = target)) +
    geom_point ( data = data_beta_E, aes (x = start, y = mean_beta, color = target)) +
    geom_point ( data = data_beta_NE, aes (x = start, y = mean_beta, color = target)) +
    labs (color = NULL) +
    ylim(0,1)+
    theme (legend.pos = "top",
           axis.title.x = element_blank(),
           axis.title.y = element_blank(),
           panel.grid.major = element_blank(),
           panel.grid.minor = element_blank(),
           text = element_text(size = 14),
           panel.border = element_blank())
  
  # plot_beta
  
  
  # mdiff
  m_gen_groups <-  gen_df %>%
    group_by(CG, target) %>% 
    summarise (mean_beta = mean (beta), start = mean (start)) %>% 
    as.data.frame() %>% split (., .$target)
  
  
  data_mdiff <- m_gen_groups$Exposed %>%
                mutate (mdiff = (mean_beta - m_gen_groups$`Non-exposed`$mean_beta))
                          # mutate (mdiff = (mean_beta * 100 /m_gen_groups$`Non-exposed`$mean_beta)-100)
  
  data_top_var <- data_mdiff %>%
                  filter (CG %in% top_var) %>%
                  mutate (top_var = "Top var")
  
  plot_mdiff <- ggplot (data = data_mdiff,
                        aes (x = start, y = mdiff)) + 
    #geom_point(aes (color = target)) + 
    geom_line (aes (x = start, y = mdiff)) +
    geom_point (aes (x = start, y = mdiff)) +
    geom_text_repel(data = data_top_var, 
                    aes (x = start, y = mdiff, label = CG),
                    box.padding = 1,
                    direction = "x", 
                    color = "indianred1") +
    geom_point (data = data_top_var,
                aes (x = start, y = mdiff ,  color = top_var))+
    geom_hline( yintercept = 0, linetype = "dashed", color = "blue", alpha = 1)+
    scale_color_manual(values = c("red"))+
    labs (color = NULL)+
    theme (legend.pos = "top", 
           axis.title.x = element_blank(), 
           axis.title.y = element_blank(), 
           panel.grid.major = element_blank(), 
           panel.grid.minor = element_blank(),
           text = element_text(size = 14), 
           panel.border = element_blank())
    
  
  # plot_mdiff
  
  # Plot gene
  plot_gen <- ggplot() + geom_alignment(TxDb.Hsapiens.UCSC.hg19.knownGene, which = genesymbol[gene], stat = "identity") + 
  theme(panel.grid.major = element_blank(), 
        panel.grid.minor = element_blank(),
        text = element_text(size = 12),
        panel.border = element_blank())
  
  # plot_gen
  
  
  
  #Track combination: 
  plotobj <- ggbio::tracks ("Gene" = plot_gen,
                            "Beta values" =  plot_beta,
                            "Difference Beta values (E - NE)" = plot_mdiff,
                            heights = c(2, 5, 5), 
                            main = gene)
  
  return (plotobj) 

}




```





Plot of the top genes in the top variables. 

```{r plots of the genes}


top_genes_dir <- file.path (out_dir, "top_genes")
# Check if output folder exist
if (!dir.exists(top_genes_dir)) {
  dir.create(top_genes_dir)
} 


plotobj <- dmg_plot_continuous(grset, beta_values, samples_metadata_fil, gene = "SDK1", top_var = top_var_limma)



for (top_gene in comp_top_genes_nz_df$Gene[1:30]){
  
  plotobj <- dmg_plot_continuous(grset, beta_values, samples_metadata_fil, gene = top_gene, top_var_limma)
  CairoPNG( width = 1500,
            height = 2000,
            file = paste0 (top_genes_dir,"/", top_gene, ".png"),
            bg = "white", units = "px", dpi = 300)
  print (plotobj)
  dev.off()
}

ls()
```





## 3. Over-Representation analysis

```{r, libraries}
library ("fgsea") 
library ("clusterProfiler") 
library ("org.Hs.eg.db")
```




Array with genes from the top_proves: 

```{r}
top_genes <-  lapply(strsplit(comp_top$UCSC_RefGene_Accession, ";"), unique)
top_genes <- unlist( top_genes[which(lapply(top_genes, length) !=0)]) %>% unique()
```



### -GO

```{r}

#Selection of the ENSEML ids
ensembl_ids <- AnnotationDbi::select(org.Hs.eg.db, 
                                     keys = top_genes, 
                                     columns = "ENSEMBL",
                                     keytype = "REFSEQ")



enrichment <- clusterProfiler::enrichGO(gene = ensembl_ids$ENSEMBL,
                                        OrgDb = org.Hs.eg.db,
                                        keyType = 'ENSEMBL',
                                        ont = "BP", 
                                        pvalueCutoff = 0.05, 
                                        qvalueCutoff = 0.02,
                                        readable = TRUE)



```



```{r}
library("enrichplot")

enrichplot::upsetplot(enrichment)

barplot(enrichment, 
        drop = TRUE, 
        showCategory = 10, 
        title = "GO Biological Pathways",
        font.size = 8)

dotplot(enrichment, font.size = 10, label_format = 60 )

ggsave ("dotplot_GO_BP_MEvPF.png", path = out_dir, units = "in", width = 9, height = 5)


goplot(enrichment, showCategory = 10)





CairoPNG( width = 2500,
          height = 2500,
          file = paste0 ("OUTPUT/top_methylation_features/", "goplot_EvPF_limma", ".png"),
          bg = "white", units = "px", dpi = 200)
goplot(enrichment, showCategory = 10)
dev.off()
```

